File: packages/cli/src/credentials.ts
import { homedir } from "node:os";
import { join } from "node:path";
import { TerminalLogger } from "@talosjs/logger";
import { LOG_OPTIONS, toYaml } from "./utils";

/**
 * Persist a credentials profile as block-style YAML under
 * `~/.talos/credentials/<fileName>`, logging the destination unless silent.
 */
export const saveCredentials = async (
  fileName: string,
  label: string,
  profile: Record<string, string>,
  silent?: boolean,
): Promise<string> => {
  const credentials = { profiles: { default: profile } };
  const credentialsPath = join(homedir(), ".talos", "credentials", fileName);
  await Bun.write(credentialsPath, `${toYaml(credentials)}\n`);

  if (!silent) {
    new TerminalLogger().success(`${label} credentials saved to ${credentialsPath}`, undefined, LOG_OPTIONS);
  }

  return credentialsPath;
};

/**
 * Read the `default` profile stored in `~/.talos/credentials/<fileName>`,
 * returning null when the file does not exist.
 */
export const readCredentials = async (fileName: string): Promise<Record<string, string> | null> => {
  const credentialsPath = join(homedir(), ".talos", "credentials", fileName);
  const file = Bun.file(credentialsPath);

  if (!(await file.exists())) {
    return null;
  }

  const parsed = Bun.YAML.parse(await file.text()) as {
    profiles?: { default?: Record<string, string> };
  };
  return parsed?.profiles?.default ?? null;
};

File: packages/rust-cli/src/utils/credentials.rs
use std::fs;
use std::path::PathBuf;

use super::yaml::{credentials_to_yaml, parse_default_profile};

fn credentials_path(file_name: &str) -> Option<PathBuf> {
    let home = dirs_home()?;
    Some(home.join(".talos").join("credentials").join(file_name))
}

fn dirs_home() -> Option<PathBuf> {
    std::env::var_os("HOME").map(PathBuf::from)
}

pub fn save_credentials(
    file_name: &str,
    label: &str,
    profile: &[(String, String)],
    silent: bool,
) -> Option<PathBuf> {
    let path = credentials_path(file_name)?;
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).ok()?;
    }
    fs::write(&path, credentials_to_yaml(profile)).ok()?;

    if !silent {
        super::style::success(format!("{label} credentials saved to {}", path.display()));
    }

    Some(path)
}

#[allow(
    dead_code,
    reason = "read side used by future consumers of saved credentials"
)]
pub fn read_credentials(file_name: &str) -> Option<Vec<(String, String)>> {
    let path = credentials_path(file_name)?;
    let content = fs::read_to_string(path).ok()?;
    Some(parse_default_profile(&content))
}
